sound/soc/sof/intel/cnl.c

```diff
@@ -150,14 +151,53 @@ static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev)
 				CNL_DSP_REG_HIPCCTL_DONE);
 }
 
+static bool cnl_is_ipc_compact(struct snd_sof_ipc_msg *msg)
+{
+	/* the compact IPC subset might be extended later */
+	if (msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE))
+		return true;
+	else
+		return false;
+}
+
+static void cnl_compact_ipc_assemble(struct snd_sof_ipc_msg *msg,
+				     u32 *dr, u32 *dd)
+{
+	struct sof_ipc_pm_gate *pm_gate;
+
+	if (msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE)) {
+		pm_gate = msg->msg_data;
+		/* set cAVS SET_D0IX IPC to the primary register */
+		*dr = CAVS_IPC_MSG_TGT | CAVS_IPC_MOD_SET_D0IX;
+		/* set the extended data register with the flag bits */
+		*dd = pm_gate->flags & CAVS_IPC_MOD_SETD0IX_BIT_MASK;
+	}
+}
+
 static int cnl_ipc_send_msg(struct snd_sof_dev *sdev,
 			    struct snd_sof_ipc_msg *msg)
 {
-	/* send the message */
-	sof_mailbox_write(sdev, sdev->host_box.offset, msg->msg_data,
-			  msg->msg_size);
-	snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
-			  CNL_DSP_REG_HIPCIDR_BUSY);
+	u32 dr = 0;
+	u32 dd = 0;
+
+	if (cnl_is_ipc_compact(msg)) {
+		/* send the message via IPC registers only */
+
+		/* first, assemble the IPC registers */
+		cnl_compact_ipc_assemble(msg, &dr, &dd);
```

Comment:
assemble is an odd wording.

'compress' would be better, it explains that we only transmit the IPC in a compressed form.

Reply:
OK.